py3status/modules/speedtest.py

```diff
@@ -0,0 +1,313 @@
+
+    def _is_running(self):
+        try:
+            self.py3.command_output(["pgrep", "speedtest-cli"])
+            return True
+        except self.py3.CommandError:
+            return False
+
+    def _get_speedtest_data(self):
+        return loads(self.py3.command_output(self.command)) or None
+
+    def _start(self):
+        self.cached_until = self.py3.time_in(self.sleep_timeout)
+
+        if not self._is_running():
+            self.cached_until = self.py3.CACHE_FOREVER
+            previous_data = self.py3.storage_get("speedtest_data") or {}
+            current_data = self._get_speedtest_data() or {}
+
+            if current_data and len(current_data) > 1:
+                # create a "total" for know if cnx is better or not
+                # between two run
+                current_data["total"] = int(current_data.get("download", 0)) + int(
+                    current_data.get("upload", 0)
+                )
+
+                # create "quality" #maybe bad name
+                if previous_data and "total" in previous_data:
+                    if current_data["total"] >= previous_data["total"]:
+                        quality_key = 4
+                    else:
+                        quality_key = 3
+                else:
+                    if current_data["total"] <= 0:
+                        quality_key = 1
+                    else:
+                        quality_key = 2
+                current_data["quality"] = quality_key
+                current_data["quality_name"] = self.quality[quality_key]
+
+        # zero-ing if not fetched, raw version and units convertion
+        for x in ["download", "upload", "bytes_received", "bytes_sent"]:
+            unit = self.unit_size if "bytes" in x else self.unit_bitrate
+            current_data[x] = current_data.get(x, 0)
+            current_data[x + "_raw"] = current_data[x]
+            current_data[x], current_data[x + "_unit"] = self.py3.format_units(
+                current_data[x], unit=unit, si=self.si_units
+            )
+
+        # extra data, not sure we want to expose
+        for x in ["server", "client"]:
+            if x in current_data:
+                for y in current_data[x]:
+                    current_data[x + "_" + y] = current_data[x][y]
+                del current_data[x]
+
+        # store last data fetched
+        self.py3.storage_set("speedtest_data", current_data)
+
+        # get speedtest result url
+        self.url = current_data.get("share")
+
+        # create placeholders
+        self.speedtest_data.update(current_data)
+        if previous_data:
+            self.speedtest_data.update(
+                {"previous_" + k: v for (k, v) in previous_data.items()}
+            )
+
+        # # cast
+        self.speedtest_data.update(
+            {
+                k: float(v)
+                for (k, v) in self.speedtest_data.items()
+                if isinstance(v, Number)
```

Comment:
Some are `int`. Users won't use this much, but we shouldn't see `0.0` or `2321.0` on the bar though.

Reply:
so I int all Number?

Reply:
Which placeholders specifically do you require to have `int/float`?

`speedtest-cli --json` gave me json with several floating numbers. There are some string ints/floats (eg two different `lat` + `lon`, `server_id`, etc). Are you casting this because you knew about https://github.com/ultrabug/py3status/issues/1391?

The only thing that can require casting is... `isprating`, `rating`, `ispdlavg`, `ispulavg`, `loggedin`. I'm not sure if users would use this to compare something with.